src/webserver.cpp

```diff
@@ -454,13 +458,49 @@ MHD_Result webserver::post_iterator(void *cls, enum MHD_ValueKind kind,
         const char *transfer_encoding, const char *data, uint64_t off, size_t size) {
     // Parameter needed to respect MHD interface, but not needed here.
     std::ignore = kind;
-    std::ignore = filename;
     std::ignore = content_type;
     std::ignore = transfer_encoding;
     std::ignore = off;
 
     struct details::modded_request* mr = (struct details::modded_request*) cls;
-    mr->dhr->set_arg(key, mr->dhr->get_arg(key) + std::string(data, size));
+
+    if (filename == nullptr || mr->ws->file_upload_target != FILE_UPLOAD_DISK_ONLY) {
+        mr->dhr->set_arg(key, mr->dhr->get_arg(key) + std::string(data, size));
+    }
+
+    if (filename && mr->ws->file_upload_target != FILE_UPLOAD_MEMORY_ONLY) {
+        // either get the existing file info struct or create a new one in the file map
+        file_info_s &file_info = mr->dhr->get_or_create_file_info(filename);
+        // if the file_system_file_name is not filled yet, this is a new entry and the name has to be set
+        // (either random or copy of the original filename)
+        if (file_info.file_system_file_name.empty()) {
+            if (mr->ws->generate_random_filename_on_upload) {
+                file_info.file_system_file_name = http_utils::generate_random_upload_filename(mr->ws->post_upload_dir);
+            }
+            else {
+                file_info.file_system_file_name = mr->ws->post_upload_dir + "/" + std::string(filename);
+            }
+            /* to not append to an already existing file, delete an already existing file the file */
+            unlink(file_info.file_system_file_name.c_str());
+        }
+
+        // if multiple files are uploaded, a different filename indicates the start of a new file, so close the previous one
+        if (!mr->upload_filename.empty() && 0 != strcmp(filename, mr->upload_filename.c_str())) {
+            mr->upload_ostrm.close();
+        }
+
+        if (!mr->upload_ostrm.is_open()) {
+            mr->upload_filename = filename;
```

Comment:
mr->upload_filename is only used internally to keep track of the currently used file

The post_iterator is called multiple times for each file (in chunks of 32 k). The currently used filename and filestream are stored in the modded_request to avoid closing and reopening the very same file for every 32 k (could decrease the upload performance). If the first file is finished and the second is started, mr->upload_filename will be overwritten to the second (or following) file.

mr->upload_filename is not meant to be used outside the post iterator but needs to be somewhere, where we can hand this to the post_iterator.

So you are correct, after the post iteration is complete the mr->upload_filename will have the name of the last uploaded file.
But once the post iteration is finished you should only access the files (and the according paths and infos) via the files map.

Reply:
Makes sense.